guava/hw10-workHibernate/src/main/java/ru/otus/hibernate/dao/UserDaoHibernate.java

```diff
@@ -0,0 +1,85 @@
+package ru.otus.hibernate.dao;
+
+
+import org.hibernate.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ru.otus.core.dao.UserDao;
+import ru.otus.core.dao.UserDaoException;
+import ru.otus.core.model.User;
+import ru.otus.core.sessionmanager.SessionManager;
+import ru.otus.hibernate.sessionmanager.DatabaseSessionHibernate;
+import ru.otus.hibernate.sessionmanager.SessionManagerHibernate;
+
+import java.util.Optional;
+
+public class UserDaoHibernate implements UserDao {
+    private static Logger logger = LoggerFactory.getLogger(UserDaoHibernate.class);
+
+    private final SessionManagerHibernate sessionManager;
+
+    public UserDaoHibernate(SessionManagerHibernate sessionManager) {
+        this.sessionManager = sessionManager;
+    }
+
+
+    @Override
+    public Optional<User> findById(long id) {
+        DatabaseSessionHibernate currentSession = sessionManager.getCurrentSession();
+        try {
+            return Optional.ofNullable(currentSession.getHibernateSession().find(User.class, id));
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public long insertUser(User user) {
+        DatabaseSessionHibernate currentSession = sessionManager.getCurrentSession();
+        try {
+            Session hibernateSession = currentSession.getHibernateSession();
+            hibernateSession.persist(user);
+            hibernateSession.flush();
+            return user.getId();
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+            throw new UserDaoException(e);
+        }
+    }
+
+    @Override
+    public void updateUser(User user) {
+        DatabaseSessionHibernate currentSession = sessionManager.getCurrentSession();
+        try {
+            Session hibernateSession = currentSession.getHibernateSession();
+            hibernateSession.merge(user);
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+            throw new UserDaoException(e);
+        }
+    }
+
+    @Override
+    public void insertOrUpdate(User user) {
+        DatabaseSessionHibernate currentSession = sessionManager.getCurrentSession();
+        try {
+            Session hibernateSession = currentSession.getHibernateSession();
+            if (user.getId() > 0) {
+                hibernateSession.merge(user);
+            } else {
+                hibernateSession.persist(user);
+                hibernateSession.flush();
+            }
```

Comment:
flush здесь лишнее, вы комитите транзакцию уровнем выше.

И это можно заменить на `hibernateSession.saveOrUpdate(user);`
```suggestion
            hibernateSession.saveOrUpdate(user);
```